doc/adr/0005-wego-core-auth-strategy.md

```diff
@@ -0,0 +1,125 @@
+
+For this reason, we want to utilize the more short-lived OAuth tokens to avoid exposing a personal access token. We also don't want to persist these tokens longer than we have to.
+
+In addition, flows that require a `CLIENT_SECRET` will require us to embed it in the binary, which leaves us open to exposing it via decompilation.
+
+### UX Concerns
+
+Given that the Weave GitOps UI and API server may be running on a user's cluster, we may not always know the OAuth callback URL ahead of time. For this reason, we will need to use OAuth flows that do not require a static `redirect_uri`. The flows listed above in the Git Providers section fit that constraint.
+
+We will need to ensure that the `redirect_uri` is configurable by the user via environment variables. This will allow users to run and expose the UI/API servers on whichever endpoint they choose.
+
+## Design
+
+Since most modern Git Providers will support a form of browser-based OAuth, **we will utilize the browser for both the CLI and UI authentication with Git Providers**.
+
+In the case of the CLI, we can utilize a short-lived browser session that will recieve the OAuth callback and pass the token to the CLI.
+
+**Note that we can skip the OAuth portion of the flow if a user provides a `--token=<some value>` flag to the CLI command:**
+
+![CLI Auth Diagram](cli_auth.svg)
+
+UI auth will work in a similar way, with a more straight-forward set of steps:
+
+![UI Auth Diagram](ui_auth.svg)
+
+TLDR:
+
+1. The user does OAuth with their Git Provider
+2. We use the resulting access_token to push a Deploy Key to the repository (giving us `git` permissions)
+3. We create a pull request for the repository via the Git Provider HTTP API
+
+For browser security, we will convert the Git Provider OAuth token to a JSON Web Token (JWT) to protect against Cross-site Scripting (XSS) attacks. The encrypted JWT will allow a malicious script to authenticate with the Weave GitOps API only, whereas passing the unencrypted OAuth token to the browser would allow a malicious script to authenticate with the Github API.
+
+Additionally, we do not plan on adding third-party scripts to the Weave GitOps UI to minimize the surface area for XSS attacks. This does NOT, however, account for NPM modules or other dependencies that we add to our app at build time.
+
+CI will inject any sensitive information into the binary to avoid keeping data in version control in a public repo. We are still susceptible to leaking data via decompilation.
+
+## Prior Art
+
+This authentication approach is inspired by other CLI tools that have very smooth user experiences. For example, the `gcloud` CLI for Google Cloud Platform utilizes the browser to authenciate the user:
+
+```
+$ gcloud auth login
+
+Your browser has been opened to visit:
+
+https://accounts.google.com/o/oauth2/auth?response_type=code&client_id=1234.apps.googleusercontent.com&redirect_uri=http%3A%2F%2Flocalhost%3A8085%2F&scope=...
+
+Opening in existing browser session.
+
```

Comment:
I should point out that the `gcloud` client code provides their `CLIENT_ID` and `SECRET` in the Python code.

_lib/googlecloudsdk/core/config.py_
```python
# TODO(b/35848109): Distribute a clientsecrets.json to avoid putting it in code.
CLOUDSDK_CLIENT_ID = '<redacted>.apps.googleusercontent.com'
CLOUDSDK_CLIENT_NOTSOSECRET = '<redacted>'
```